flink-ml-lib/src/main/java/org/apache/flink/ml/common/broadcast/BroadcastContext.java

```diff
@@ -18,106 +18,54 @@
-     * in an atomic operation".
-     */
-    private static ReentrantReadWriteLock lock = new ReentrantReadWriteLock();
+    private static final Map<String, Tuple2<Boolean, List<?>>> BROADCAST_VARIABLES =
+            new ConcurrentHashMap<>();
 
-    public static void putBroadcastVariable(
-            Tuple2<String, Integer> key, Tuple2<Boolean, List<?>> variable) {
-        lock.writeLock().lock();
-        try {
-            broadcastVariables.put(key, variable);
-        } finally {
-            lock.writeLock().unlock();
-        }
+    @VisibleForTesting
+    public static void putBroadcastVariable(String key, Tuple2<Boolean, List<?>> variable) {
+        BROADCAST_VARIABLES.put(key, variable);
     }
 
     /**
-     * get the cached list with the given key.
-     *
-     * @param key
-     * @param <T>
-     * @return the cache list.
-     */
-    public static <T> List<T> getBroadcastVariable(Tuple2<String, Integer> key) {
-        lock.readLock().lock();
-        List<?> result = null;
-        try {
-            result = broadcastVariables.get(key).f1;
-        } finally {
-            lock.readLock().unlock();
-        }
-        return (List<T>) result;
-    }
-
-    /**
-     * get broadcast variables by name
+     * gets broadcast variables by name if this broadcast variable is fully cached.
      *
      * @param name
      * @param <T>
-     * @return
+     * @return the cache broadcast variable. Return null if it is not fully cached.
      */
+    @VisibleForTesting
     public static <T> List<T> getBroadcastVariable(String name) {
-        lock.readLock().lock();
-        List<?> result = null;
-        try {
-            for (Tuple2<String, Integer> nameAndPartitionId : broadcastVariables.keySet()) {
-                if (name.equals(nameAndPartitionId.f0) && isCacheFinished(nameAndPartitionId)) {
-                    result = broadcastVariables.get(nameAndPartitionId).f1;
-                    break;
-                }
-            }
-        } finally {
-            lock.readLock().unlock();
+        Tuple2<Boolean, List<?>> cacheReadyAndList = BROADCAST_VARIABLES.get(name);
+        if (cacheReadyAndList.f0) {
+            return (List<T>) cacheReadyAndList.f1;
         }
-        return (List<T>) result;
+        return null;
     }
 
-    public static void remove(Tuple2<String, Integer> key) {
-        lock.writeLock().lock();
-        try {
-            broadcastVariables.remove(key);
-        } finally {
-            lock.writeLock().unlock();
-        }
+    @VisibleForTesting
+    public static void remove(String key) {
+        BROADCAST_VARIABLES.remove(key);
     }
 
-    public static void markCacheFinished(Tuple2<String, Integer> key) {
-        lock.writeLock().lock();
-        try {
-            broadcastVariables.get(key).f0 = true;
-        } finally {
-            lock.writeLock().unlock();
-        }
+    @VisibleForTesting
+    public static void markCacheFinished(String key) {
```

Comment:
We should need to explicitly notify the wrapper operator by emitting one mail? otherwise the wrapper operator may stalled in the `endInputX`.